File: typescript/src/crypto/sign.ts
import { sign as signMessage, verify as verifySignature } from 'stedy'
import { importPrivateSignKey, importPublicSignKey } from '../utils'

export const sign = async (
  ourPrivateKey: BufferSource,
  message: BufferSource
) => signMessage(await importPrivateSignKey(ourPrivateKey), message)

export const verify = async (
  message: BufferSource,
  publicKey: BufferSource,
  signature: BufferSource
) => verifySignature(message, await importPublicSignKey(publicKey), signature)

File: cplusplus/src/crypto/sign.cpp
#include "crypto.hpp"
#include "sodium.h"

namespace autograph {

bool sign(unsigned char *signature, const unsigned char *private_key,
          const unsigned char *message, const unsigned long long message_size) {
  return crypto_sign_detached(signature, nullptr, message, message_size,
                              private_key) == 0;
}

bool verify(const unsigned char *public_key, const unsigned char *message,
            const unsigned long long message_size,
            const unsigned char *signature) {
  return crypto_sign_verify_detached(signature, message, message_size,
                                     public_key) == 0;
}

}  // namespace autograph
